### PR TITLE
Introduce retry

### PR DESCRIPTION
**Pull request by xitronix**

### CODE REVIEW

File: universal-login-relayer/lib/services/messages/MessageQueueStore.ts [hunk 1]
```diff
@@ -57,6 +57,15 @@ export default class MessageQueueStore implements IMessageQueueStore {
       .first();
   }
 
+  async getStatus(id: string) {
+    return this.database(this.tableName)
+    .first()
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/services/messages/MessageQueueStore.ts]: `\t`

File: universal-login-relayer/lib/services/messages/MessageQueueStore.ts [hunk 2]
```diff
@@ -57,6 +57,15 @@ export default class MessageQueueStore implements IMessageQueueStore {
       .first();
   }
 
+  async getStatus(id: string) {
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/services/messages/MessageQueueStore.ts]: `getStatusIfMined`

File: universal-login-commons/lib/types/message.ts [hunk 3]
```diff
@@ -22,10 +22,16 @@ export type UnsignedMessage = Omit<SignedMessage, 'signature'>;
 
 export type MessageStatus = {
   id?: string,
+  error?: string,
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/types/message.ts]: `MessageStatus.ts`:

```ts
function isExecuted(MessageStatus message) : boolean {
  return message.id || message.error;
}
```

File: universal-login-relayer/lib/services/messages/MessageQueueStore.ts [hunk 4]
```diff
@@ -57,6 +57,15 @@ export default class MessageQueueStore implements IMessageQueueStore {
...
+    return this.database(this.tableName)
+    .first()
+    .where('id', id)
+    .whereNotNull('hash')
+    .orWhereNotNull('error')
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/services/messages/MessageQueueStore.ts]: remove

File: universal-login-relayer/test/helpers/MessageQueueMemoryStore.ts [hunk 5]
```diff
@@ -29,6 +29,14 @@ export default class MessageQueueMemoryStore implements IMessageQueueStore {
     return this.messageEntries[0];
   }
 
+  async getStatus(id: string) {
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/helpers/MessageQueueMemoryStore.ts]: `getStatusById`

File: universal-login-relayer/lib/services/messages/MessageQueueStore.ts (hunk 2 above)
Comment (marekkirejczyk) [reply universal-login-relayer/lib/services/messages/MessageQueueStore.ts]:
```ts
class MessageHandler {
...
  getMessageStatusById(id: string) : MessageStatus {
    const message : MessageQueueStatus = messageQueue.getMessageStatusById(id);
    return QueueStatusToMessageStatus(message);
  }
}
```

File: universal-login-sdk/lib/sdk.ts [hunk 6]
```diff
@@ -127,17 +128,12 @@ class UniversalLoginSDK {
...
-    let status;
-    do {
-      if (elapsed > timeout) {
-        throw Error('Timeout');
-      }
-      await sleep(tick);
-      elapsed += tick;
-      status =  await this.relayerApi.getStatusById(messageId);
-    } while (status && !status.transactionHash && !status.error);
+  private isNotExecuted (messageStatus: MessageStatus){
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]: `isNotExecuted` -> `isExecuted`

File: universal-login-sdk/lib/sdk.ts [hunk 7]
```diff
@@ -127,17 +128,12 @@ class UniversalLoginSDK {
...
+    return !messageStatus.transactionHash && !messageStatus.error;
+  }
+
+  async waitForStatus(messageHash: string) {
+    const status = await retry(async () => this.relayerApi.getStatus(messageHash), this.isNotExecuted);
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]: (message) => !this.isExecuted(message)

File: universal-login-relayer/lib/services/MessageHandler.ts [hunk 8]
```diff
@@ -54,24 +55,39 @@ class MessageHandler {
     return this.authorisationService.removeRequest(message.from, key);
   }
 
-  async getStatus(hash: string) {
-    if (await this.pendingMessages.isPresent(hash)){
-      return this.pendingMessages.getStatus(hash);
+  async getStatus(messageHash: string) {
+    if (await this.pendingMessages.isPresent(messageHash)){
+      const pendingStatus = await this.pendingMessages.getStatus(messageHash);
+      const queueStatus = await this.messageQueue.getStatus(messageHash);
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/services/MessageHandler.ts]: || {};

File: universal-login-sdk/lib/SdkSigner.ts [hunk 9]
```diff
@@ -43,6 +44,7 @@ export class SdkSigner extends ethers.Signer {
       message.value = await transaction.value;
     }
     const {transactionHash} = await this.sdk.execute(message, this.wallet.privateKey);
-    return this.provider.getTransaction(transactionHash);
+    ensure(!!transactionHash, Error, 'Transaction hash is not found in Message Status');
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/SdkSigner.ts]: `ensureNotFalsy(transactionHash, Error, 'Transaction hash is not found in Message Status');`

File: universal-login-sdk/lib/utils/retry.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
+import {now} from 'moment';
+import {BasicFunction, ensure, sleep} from '@universal-login/commons';
+
+export async function retry(
+  callback: BasicFunction,
+  retryWhile : (result?: ReturnType<typeof callback>) => boolean = (result) => !result,
+  timeout: number = 5000,
+  tick: number = 1000
+) {
+  const stopTime = now() + timeout;
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/utils/retry.ts]: `Date.now()`

File: universal-login-relayer/lib/services/MessageHandler.ts [hunk 11]
```diff
@@ -54,24 +55,39 @@ class MessageHandler {
     return this.authorisationService.removeRequest(message.from, key);
   }
 
-  async getStatus(hash: string) {
-    if (await this.pendingMessages.isPresent(hash)){
-      return this.pendingMessages.getStatus(hash);
+  async getStatus(messageHash: string) {
+    if (await this.pendingMessages.isPresent(messageHash)){
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/services/MessageHandler.ts]:
```ts
if (!await this.pendingMessages.isPresent(messageHash)) {
  return null;
}
```

File: universal-login-relayer/lib/services/MessageHandler.ts [hunk 12]
```diff
@@ -54,24 +55,39 @@ class MessageHandler {
...
+      const pendingStatus = await this.pendingMessages.getStatus(messageHash);
+      const queueStatus = await this.messageQueue.getStatus(messageHash);
+      if (queueStatus) {
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/services/MessageHandler.ts]: Unnecessary

File: universal-login-relayer/lib/services/messages/PendingMessages.ts [hunk 13]
```diff
@@ -77,7 +77,7 @@ export default class PendingMessages {
 
   async ensureCorrectExecution(messageHash: string) {
     const {required, transactionHash, totalCollected} = await this.messagesStore.getStatus(messageHash, this.wallet);
-    ensure(transactionHash === '0x0', DuplicatedExecution);
+    ensure(!transactionHash, DuplicatedExecution);
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/services/messages/PendingMessages.ts]: `ensureNotNull`

File: universal-login-relayer/migrations/20190613150800_alter_table_transactions.js [hunk 14]
```diff
@@ -0,0 +1,29 @@
+/*
+  THIS IS A MIGRATION
+  This file SHOULD NOT be RENAMED or CHANGED in ANY WAY!
+  If you need to change the database schema create a new migration.
+*/
+
+exports.up = async (knex) => {
+  await knex.schema.renameTable('transactions', 'finalMessages');
+  await knex.schema.alterTable('finalMessages', (table) => {
+    table.dropColumn('hash');
```
Comment (marekkirejczyk) [inline universal-login-relayer/migrations/20190613150800_alter_table_transactions.js]: rename, alter

File: universal-login-relayer/test/helpers/MessageQueueMemoryStore.ts [hunk 15]
```diff
@@ -1,39 +1,45 @@
...
 
   async add (signedMessage: SignedMessage) {
-    this.counter++;
+    const messageHash = calculateMessageHash(signedMessage);
     this.messageEntries.push({
       message: signedMessage,
-      id: this.counter.toString(),
-      hash: 'hash',
+      messageHash,
+      transactionHash: 'transactionHash',
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/helpers/MessageQueueMemoryStore.ts]: ??

File: universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts [hunk 16]
```diff
@@ -2,15 +2,17 @@ import {expect} from 'chai';
...
 import {getTestSignedMessage} from '../../../config/message';
 
 describe('INT: Message Queue Store', async () => {
   let messageQueueStore: MessageQueueStore;
   let signedMessage: SignedMessage;
+  let messageHash: string;
 
   before(async () => {
     signedMessage = await getTestSignedMessage();
+    messageHash = calculateMessageHash(signedMessage);
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts]: `expectedMessageHash`

File: universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts [hunk 17]
```diff
@@ -23,14 +25,14 @@ describe('INT: Message Queue Store', async () => {
   });
 
   it('add message', async () =>  {
-    const idFirst = await messageQueueStore.add(signedMessage);
-    expect(idFirst).to.be.a('number');
-    expect(idFirst).to.be.at.least(1);
+    const hash = await messageQueueStore.add(signedMessage);
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts]: `messageHash`

File: universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts [hunk 18]
```diff
@@ -23,14 +25,14 @@ describe('INT: Message Queue Store', async () => {
...
+    expect(hash).to.be.a('string');
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts]: Unnecessary

File: universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts [hunk 19]
```diff
@@ -42,24 +44,24 @@ describe('INT: Message Queue Store', async () => {
...
-    await messageQueueStore.markAsSuccess(idFirst, '0x000');
-    const id2 = (await messageQueueStore.getNext()).id;
-    expect(id2).to.be.equal(idSecond);
-    expect(id2).to.be.a('number');
-    expect(id2).to.be.at.least(2);
-    await messageQueueStore.markAsSuccess(idSecond, '0x000');
+    const messageHash2 = await messageQueueStore.add(signedMessage2);
+    const nextMessageHash = (await messageQueueStore.getNext()).messageHash;
+    expect(nextMessageHash).to.be.equal(messageHash1);
+    expect(nextMessageHash).to.be.a('string');
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts]: Unnecessary

File: universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts [hunk 20]
```diff
@@ -42,24 +44,24 @@ describe('INT: Message Queue Store', async () => {
...
+    expect(nextMessageHash).to.be.eq(messageHash);
+    await messageQueueStore.markAsSuccess(messageHash1, '0x000');
+    const nextMessageHash2 = (await messageQueueStore.getNext()).messageHash;
+    expect(nextMessageHash2).to.be.equal(messageHash2);
+    expect(nextMessageHash2).to.be.a('string');
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts]: Unnecessary

File: universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts [hunk 21]
```diff
@@ -42,24 +44,24 @@ describe('INT: Message Queue Store', async () => {
...
-    expect(id2).to.be.equal(idSecond);
-    expect(id2).to.be.a('number');
-    expect(id2).to.be.at.least(2);
-    await messageQueueStore.markAsSuccess(idSecond, '0x000');
+    const messageHash2 = await messageQueueStore.add(signedMessage2);
+    const nextMessageHash = (await messageQueueStore.getNext()).messageHash;
+    expect(nextMessageHash).to.be.equal(messageHash1);
+    expect(nextMessageHash).to.be.a('string');
+    expect(nextMessageHash).to.be.eq(messageHash);
+    await messageQueueStore.markAsSuccess(messageHash1, '0x000');
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts]: '0x000' -> extract meaningful constant with meaningful value

File: universal-login-relayer/test/unit/MessageQueue.ts [hunk 22]
```diff
@@ -40,7 +40,7 @@ describe('UNIT: Message Queue Service', async () => {
     await waitExpect(() => expect(executor.executeAndWait).to.be.calledOnce);
   });
 
-  it('should execute signedMessage if id added before start', async () => {
+  it('should execute signedMessage if it added before start', async () => {
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/unit/MessageQueue.ts]: should execute pending signedMessage after `start`

File: universal-login-sdk/lib/sdk.ts (hunk 6 above)
Comment (xitronix) [reply universal-login-sdk/lib/sdk.ts]: ✔️

File: universal-login-sdk/lib/sdk.ts (hunk 7 above)
Comment (xitronix) [reply universal-login-sdk/lib/sdk.ts]: ✔️

File: universal-login-relayer/lib/services/MessageHandler.ts (hunk 8 above)
Comment (xitronix) [reply universal-login-relayer/lib/services/MessageHandler.ts]: ✔️

File: universal-login-relayer/lib/services/MessageHandler.ts (hunk 11 above)
Comment (xitronix) [reply universal-login-relayer/lib/services/MessageHandler.ts]: ✔️

File: universal-login-relayer/lib/services/MessageHandler.ts (hunk 12 above)
Comment (xitronix) [reply universal-login-relayer/lib/services/MessageHandler.ts]: ✔️

File: universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts (hunk 16 above)
Comment (xitronix) [reply universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts]: ✔️

File: universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts (hunk 17 above)
Comment (xitronix) [reply universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts]: ✔️

File: universal-login-sdk/lib/utils/retry.ts (hunk 10 above)
Comment (xitronix) [reply universal-login-sdk/lib/utils/retry.ts]: ✔️

File: universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts (hunk 19 above)
Comment (xitronix) [reply universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts]: ✔️

File: universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts (hunk 20 above)
Comment (xitronix) [reply universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts]: ✔️

File: universal-login-relayer/test/unit/MessageQueue.ts (hunk 22 above)
Comment (xitronix) [reply universal-login-relayer/test/unit/MessageQueue.ts]: ✔️

File: universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts (hunk 21 above)
Comment (xitronix) [reply universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts]: ✔️

File: universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts (hunk 18 above)
Comment (xitronix) [reply universal-login-relayer/test/relayer/services/messages/MessageQueueStore.ts]: ✔️

File: universal-login-relayer/test/helpers/MessageQueueMemoryStore.ts (hunk 15 above)
Comment (xitronix) [reply universal-login-relayer/test/helpers/MessageQueueMemoryStore.ts]: ✔️

File: universal-login-sdk/lib/sdk.ts [hunk 23]
```diff
@@ -127,17 +128,14 @@ class UniversalLoginSDK {
...
-      status =  await this.relayerApi.getStatusById(messageId);
-    } while (status && !status.transactionHash && !status.error);
+  private isExecuted (messageStatus: MessageStatus){
+    return !!messageStatus.transactionHash || !!messageStatus.error;
+  }
+
+  async waitForStatus(messageHash: string) {
+    const getStatus = async () => this.relayerApi.getStatus(messageHash);
+    const isNotExecuted = (messageStatus: MessageStatus) => !this.isExecuted(messageStatus);
+    const status = await retry(getStatus, isNotExecuted);
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]: below:
`ensure(! status.error, Error, status.error)`

File: universal-login-commons/lib/constants/constants.ts [hunk 24]
```diff
@@ -18,6 +18,8 @@ export const TEST_ACCOUNT_ADDRESS = '0x0000000000000000000000000000000000000001'
 
 export const TEST_PRIVATE_KEY = '0x63f01680950dc70f2eb8f373de0c360fcbb89ef437f2f6f2f0a1797979e490a4';
 
+export const TEST_HASH = '0x0000000000000000000000000000000000000000000000000000000000000000';
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/constants/constants.ts]: Better do more random numbers

File: universal-login-sdk/lib/SdkSigner.ts (hunk 9 above)
Comment (xitronix) [reply universal-login-sdk/lib/SdkSigner.ts]: ✔️

File: universal-login-relayer/migrations/20190613150800_alter_table_transactions.js (hunk 14 above)
Comment (xitronix) [reply universal-login-relayer/migrations/20190613150800_alter_table_transactions.js]: ✔️

File: universal-login-relayer/test/helpers/MessageQueueMemoryStore.ts [hunk 25]
```diff
@@ -1,39 +1,45 @@
...
 
   async add (signedMessage: SignedMessage) {
-    this.counter++;
+    const messageHash = calculateMessageHash(signedMessage);
     this.messageEntries.push({
       message: signedMessage,
-      id: this.counter.toString(),
-      hash: 'hash',
+      messageHash,
+      transactionHash: TEST_HASH,
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/helpers/MessageQueueMemoryStore.ts]: Unnecessary

**Pull request merged**